core/embed/rust/src/trezorhal/storage_field.rs

```diff
@@ -0,0 +1,391 @@
+    fn is_set(&self) -> bool;
+    fn delete(&self) -> StorageResult<()>;
+}
+impl<T> FieldOpsBase for Field<T> {
+    fn is_set(&self) -> bool {
+        storage::has(self.appkey())
+    }
+
+    fn delete(&self) -> StorageResult<()> {
+        // If field is not there, storage::delete() would return Err()
+        if !self.is_set() {
+            return Ok(());
+        }
+
+        if !storage::delete(self.appkey()) {
+            Err(StorageError::WriteFailed)
+        } else {
+            Ok(())
+        }
+    }
+}
+
+pub trait FieldGetSet<T> {
+    fn get(&self) -> Option<T>;
+    fn set(&self, value: T) -> StorageResult<()>;
+}
+
+pub trait FieldOps<T>: FieldOpsBase + FieldGetSet<T> {}
+impl<T> FieldOps<T> for Field<T> where Field<T>: FieldGetSet<T> {}
+
+impl FieldGetSet<u32> for Field<u32> {
+    fn get(&self) -> Option<u32> {
+        let mut buf = [0u8; 4];
+        let len = storage::get(self.appkey(), &mut buf);
+        if matches!(len, Ok(4)) {
+            Some(u32::from_be_bytes(buf))
+        } else {
+            None
+        }
+    }
+
+    fn set(&self, val: u32) -> StorageResult<()> {
+        storage::set(self.appkey(), &val.to_be_bytes())
+    }
+}
+
+impl FieldGetSet<u16> for Field<u16> {
+    fn get(&self) -> Option<u16> {
+        let mut buf = [0u8; 2];
+        let len = storage::get(self.appkey(), &mut buf);
+        if matches!(len, Ok(2)) {
+            Some(u16::from_be_bytes(buf))
+        } else {
+            None
+        }
+    }
+
+    fn set(&self, val: u16) -> StorageResult<()> {
+        storage::set(self.appkey(), &val.to_be_bytes())
+    }
+}
+
+impl FieldGetSet<u8> for Field<u8> {
+    fn get(&self) -> Option<u8> {
+        let mut buf = [0u8];
+        let len = storage::get(self.appkey(), &mut buf);
+        if matches!(len, Ok(1)) {
+            Some(buf[0])
+        } else {
+            None
+        }
+    }
+
+    fn set(&self, val: u8) -> StorageResult<()> {
+        storage::set(self.appkey(), &[val])
+    }
+}
+
+impl FieldGetSet<bool> for Field<bool> {
+    fn get(&self) -> Option<bool> {
+        let mut buf = [0u8];
+        let len = storage::get(self.appkey(), &mut buf);
+        if matches!(len, Ok(1)) {
+            Some(buf[0] == TRUE_BYTE)
+        } else {
+            None
+        }
+    }
+
+    fn set(&self, val: bool) -> StorageResult<()> {
+        storage::set(self.appkey(), &[if val { TRUE_BYTE } else { FALSE_BYTE }])
+    }
+}
+
+/// Helper function defined only on boolean Fields
+pub fn set_true_or_delete(field: impl FieldOps<bool>, value: bool) -> StorageResult<()> {
+    if value {
+        field.set(true)
+    } else {
+        field.delete()
+    }
+}
+
+impl<const N: usize> FieldGetSet<String<N>> for Field<String<N>> {
+    fn get(&self) -> Option<String<N>> {
+        let mut buf = [0u8; MAX_STRING];
+        let len = storage::get(self.appkey(), &mut buf);
+        len.ok().and_then(|len| {
+            let string = match str::from_utf8(&buf[..len as usize]) {
```

Comment:
Maybe return the subslice directly from `storage::get(appkey: u16, data: &mut [u8]) -> StorageResult<&[u8]>`?